File: scripts/import_srd_spells.py
#!/usr/bin/env python3
"""
Transform SRD spells from dnd-data repository to local spells.json format.
"""

import urllib.request
import json
import re
import os

def slugify(name):
    """Convert spell name to kebab-case id."""
    # Handle specific spell name patterns
    name = name.lower()
    # Remove special characters except hyphens
    name = re.sub(r'[^a-z0-9\s-]', '', name)
    # Replace spaces with hyphens
    name = re.sub(r'\s+', '-', name)
    # Remove consecutive hyphens
    name = re.sub(r'-+', '-', name)
    return name

def parse_components(comp_str):
    """Parse component string like 'V, S, M' into structured format."""
    if not comp_str:
        return {"V": False, "S": False, "M": None}
    
    comp_str = comp_str.upper()
    result = {
        "V": "V" in comp_str or "VERBAL" in comp_str,
        "S": "S" in comp_str or "SOMATIC" in comp_str,
        "M": None
    }
    
    # Extract material component if present
    if "M" in comp_str:
        # Try to extract text after M in parentheses
        m_match = re.search(r'M\s*(?:\(([^)]+)\))?', comp_str)
        if m_match and m_match.group(1):
            result["M"] = m_match.group(1).lower()
        else:
            result["M"] = True  # Material required but no specific description
    
    return result

def extract_damage(description, level):
    """Extract damage dice and type from description."""
    damage = None

    # Pattern for damage dice: 1d4, 2d6, 10d8, etc.
    dmg_pattern = r'(\d+d\d+(?:\s*\+\s*\d+)?)\s+(acid|cold|fire|force|lightning|necrotic|poison|radiant|thunder|psychic|slashing|piercing|bludgeoning)'
    match = re.search(dmg_pattern, description, re.IGNORECASE)
    if match:
        damage = {
            "entries": [{
                "dice": match.group(1),
                "type": match.group(2).capitalize()
            }]
        }

    return damage

def extract_save(description):
    """Extract saving throw from description."""
    saves = ["Strength", "Dexterity", "Constitution", "Intelligence", "Wisdom", "Charisma"]
    for save in saves:
        if re.search(rf'\b{save}\b', description, re.IGNORECASE):
            return save
    return None

def check_attack(description):
    """Check if spell involves an attack roll."""
    return "attack" in description.lower() or "ranged spell attack" in description.lower()

def parse_duration(duration_str):
    """Parse duration string and check for concentration."""
    if not duration_str:
        return "Instantaneous", False
    
    concentration = "concentration" in duration_str.lower()
    return duration_str, concentration

def check_ritual(casting_time, description):
    """Check if spell can be cast as ritual."""
    ritual_keywords = ["ritual", "(ritual)"]
    for kw in ritual_keywords:
        if kw.lower() in casting_time.lower() or kw.lower() in description.lower():
            return True
    return False

def transform_spell(spell):
    """Transform a dnd-data spell to local format."""
    props = spell.get("properties", {})
    desc = spell.get("description", "")
    
    # Parse basic fields
    casting_time = props.get("Casting Time", "Action")
    range_str = props.get("Range", props.get("data-RangeAoe", "Self"))
    duration_str = props.get("Duration", "Instantaneous")
    components_str = props.get("Components", "")
    level = props.get("Level", 0)
    school = props.get("School", "")
    save = props.get("Save", extract_save(desc))
    
    # Clean up casting time
    if "Action" in casting_time and "bonus" not in casting_time.lower():
        casting_time = "Action"
    elif "bonus action" in casting_time.lower():
        casting_time = "Bonus Action"
    elif "reaction" in casting_time.lower():
        casting_time = "Reaction"
    elif "minute" in casting_time.lower() or "hour" in casting_time.lower():
        # Keep original format
        pass
    
    duration, concentration = parse_duration(duration_str)
    components = parse_components(components_str)
    damage = extract_damage(desc, level)
    ritual = check_ritual(casting_time, desc)
    attack = check_attack(desc)
    
    # Determine if it's a healing spell
    heal = None
    if "regains" in desc.lower() or "heal" in desc.lower():
        # Try to extract healing dice
        heal_match = re.search(r'(\d+d\d+)', desc)
        if heal_match:
            heal = {"dice": heal_match.group(1)}
    
    return {
        "id": slugify(spell.get("name", "")),
        "name": spell.get("name", ""),
        "level": level,
        "school": school.capitalize() if school else "Unknown",
        "castingTime": casting_time,
        "range": range_str,
        "components": components,
        "duration": duration,
        "concentration": concentration,
        "ritual": ritual,
        "description": desc,
        "damage": damage,
        "heal": heal,
        "save": save,
        "attack": attack,
        "source": spell.get("book", spell.get("publisher", "Unknown"))
    }

def is_srd_spell(spell):
    """Check if spell is from SRD (Wizards of the Coast)."""
    publisher = spell.get("publisher", "")
    book = spell.get("book", "")
    
    srd_sources = [
        "Wizards of the Coast",
        "Player's Handbook",
        "Free Basic Rules",
        "Essentials Kit",
        "Monster Manual",
        "Dungeon Master's Guide",
        "Player's Handbook (2024)",
        "Free Basic Rules (2024)"
    ]
    
    # Check publisher
    if "Wizards of the Coast" in publisher:
        return True
    
    # Check book for SRD indicators
    for source in srd_sources:
        if source in book:
            return True
    
    return False

def main():
    print("Fetching dnd-data spells...")
    
    # Fetch dnd-data spells
    url = 'https://raw.githubusercontent.com/nick-aschenbach/dnd-data/main/data/spells.json'
    with urllib.request.urlopen(url) as response:
        dnd_spells = json.loads(response.read().decode())
    
    print(f"Total spells in dnd-data: {len(dnd_spells)}")
    
    # Filter SRD spells
    srd_spells = [s for s in dnd_spells if is_srd_spell(s)]
    print(f"SRD spells: {len(srd_spells)}")
    
    # Transform spells
    transformed = []
    for spell in srd_spells:
        try:
            transformed_spell = transform_spell(spell)
            transformed.append(transformed_spell)
        except Exception as e:
            print(f"Error transforming {spell.get('name', 'Unknown')}: {e}")
    
    print(f"Successfully transformed: {len(transformed)} spells")
    
    # Load existing local data
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(script_dir)
    local_file = os.path.join(project_root, 'static', 'spells.json')
    existing_ids = set()
    existing_spells = []
    
    if os.path.exists(local_file):
        with open(local_file, 'r') as f:
            existing_spells = json.load(f)
        existing_ids = {s['id'] for s in existing_spells}
        print(f"Existing local spells: {len(existing_spells)}")
    
    # Merge: keep existing, add new
    existing_dict = {s['id']: s for s in existing_spells}
    merged = list(existing_spells)
    
    added_count = 0
    for spell in transformed:
        if spell['id'] not in existing_ids:
            existing_dict[spell['id']] = spell
            added_count += 1
    
    print(f"New spells to add: {added_count}")
    
    # Sort by level then name
    all_spells = list(existing_dict.values())
    all_spells.sort(key=lambda x: (x['level'], x['name']))
    
    # Save
    with open(local_file, 'w') as f:
        json.dump(all_spells, f, indent=2, ensure_ascii=False)
    
    print(f"Total spells after merge: {len(all_spells)}")
    print(f"Saved to: {local_file}")
    
    # Show breakdown by level
    level_counts = {}
    for spell in all_spells:
        lvl = spell['level']
        level_counts[lvl] = level_counts.get(lvl, 0) + 1
    
    print("\nSpells by level:")
    for lvl in sorted(level_counts.keys()):
        print(f"  Level {lvl}: {level_counts[lvl]}")

if __name__ == "__main__":
    main()

File: scripts/import_srd_spells.ts
#!/usr/bin/env node
/**
 * Transform SRD spells from dnd-data to local spells.json format.
 * Rewritten from Python to TypeScript.
 */

import { writeFileSync, readFileSync, existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

// ── Types ────────────────────────────────────────────────────────

interface DndDataSpell {
  name: string;
  properties?: { Level?: number; School?: string; 'Casting Time'?: string; Range?: string; 'data-RangeAoe'?: string; Duration?: string; Components?: string; Save?: string };
  description?: string;
  book?: string;
  publisher?: string;
  [key: string]: unknown;
}

interface CantripUpgradeEntry { atCharacterLevel: 5 | 11 | 17; damage?: { dice: string; type: string }[] }
interface SpellDamageEntry { dice: string; type: string }
interface SpellDamage { entries: SpellDamageEntry[]; additional?: SpellDamageEntry[] }
interface SpellHeal { dice: string }
interface Spell {
  id: string; name: string; level: number; school: string;
  castingTime: string; range: string; components: string[]; duration: string;
  concentration: boolean; ritual: boolean; description: string[];
  cantripUpgrade?: CantripUpgradeEntry[]; usingAHigherLevelSpellSlot?: string[];
  damage?: SpellDamage; heal?: SpellHeal; save?: string; attack?: boolean;
  source: string; classes?: string[];
}

// ── Config ───────────────────────────────────────────────────────

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const localSpellsPath = resolve(__dirname, '..', 'static', 'srd', 'spells.json');

// ── Helpers ────────────────────────────────────────────────────

function slugify(name: string): string {
  return name.toLowerCase().replace(/[^a-z0-9\s-]/g, '').replace(/\s+/g, '-').replace(/-+/g, '-').replace(/^-|-$/g, '');
}

function parseComponents(compStr: string | undefined): string[] {
  if (!compStr) return [];
  const upper = compStr.toUpperCase();
  const comps: string[] = [];
  if (upper.includes('V')) comps.push('V');
  if (upper.includes('S')) comps.push('S');
  if (upper.includes('M')) comps.push('M');
  return comps;
}

function parseCastingTime(ct: string | undefined): string {
  if (!ct) return 'Action';
  if (/bonus\s*action/i.test(ct)) return 'Bonus Action';
  if (/reaction/i.test(ct)) return 'Reaction';
  return 'Action';
}

function parseDuration(durationStr: string | undefined): { duration: string; concentration: boolean } {
  if (!durationStr) return { duration: 'Instantaneous', concentration: false };
  return { duration: durationStr, concentration: /concentration/i.test(durationStr) };
}

function checkRitual(castingTime: string, description: string): boolean {
  return /ritual/i.test(castingTime) || /ritual/i.test(description);
}

function extractSave(description: string): string | undefined {
  for (const save of ['Strength', 'Dexterity', 'Constitution', 'Intelligence', 'Wisdom', 'Charisma']) {
    if (new RegExp(`\\b${save}\\b`, 'i').test(description)) return save;
  }
  return undefined;
}

function checkAttack(description: string): boolean {
  return /attack/i.test(description) || /ranged spell attack/i.test(description);
}

function extractDamage(description: string): SpellDamage | undefined {
  const matches = [...description.matchAll(/(\d+d\d+(?:\s*\+\s*\d+)?)\s+(acid|cold|fire|force|lightning|necrotic|poison|radiant|thunder)/gi)];
  if (matches.length === 0) return undefined;
  return { entries: matches.map(m => ({ dice: m[1], type: m[2].charAt(0).toUpperCase() + m[2].slice(1).toLowerCase() })) };
}

function extractHeal(description: string): SpellHeal | undefined {
  if (!/regain|heal|hit point/i.test(description)) return undefined;
  const m = description.match(/(\d+d\d+)/);
  return m ? { dice: m[1] } : undefined;
}

// ── Description Parsing ─────────────────────────────────────────

function parseDescription(fullDescription: string): {
  description: string[]; cantripUpgrade?: CantripUpgradeEntry[]; usingAHigherLevelSpellSlot?: string[];
} {
  let desc = fullDescription.trim();

  // Extract "Cantrip Upgrade." section
  let cantripUpgradeText = '';
  const cantripIdx = desc.search(/Cantrip Upgrade/i);
  if (cantripIdx >= 0) {
    const nextIdx = desc.indexOf('Using a Higher-Level Spell Slot', cantripIdx);
    if (nextIdx >= 0) {
      cantripUpgradeText = desc.substring(cantripIdx, nextIdx).replace(/Cantrip Upgrade\.?\s*/i, '').trim();
      desc = desc.substring(0, cantripIdx) + ' ' + desc.substring(nextIdx);
    } else {
      cantripUpgradeText = desc.substring(cantripIdx).replace(/Cantrip Upgrade\.?\s*/i, '').trim();
      desc = desc.substring(0, cantripIdx);
    }
  }

  // Extract "Using a Higher-Level Spell Slot." section
  let usingAHigherLevelSpellSlotText = '';
  const higherLevelIdx = desc.search(/Using a Higher-Level Spell Slot/i);
  if (higherLevelIdx >= 0) {
    usingAHigherLevelSpellSlotText = desc.substring(higherLevelIdx).replace(/Using a Higher-Level Spell Slot\.?\s*/i, '').trim();
    desc = desc.substring(0, higherLevelIdx);
  }

  desc = desc.replace(/\s+/g, ' ').trim();

  const mainDescription = desc.split(/\n\n+/).map(p => p.trim()).filter(p => p.length > 0);
  const result: { description: string[]; cantripUpgrade?: CantripUpgradeEntry[]; usingAHigherLevelSpellSlot?: string[] } = {
    description: mainDescription.length > 0 ? mainDescription : (desc ? [desc] : [])
  };

  if (cantripUpgradeText) result.cantripUpgrade = parseCantripUpgrade(cantripUpgradeText);
  if (usingAHigherLevelSpellSlotText) result.usingAHigherLevelSpellSlot = [usingAHigherLevelSpellSlotText];

  return result;
}

function parseCantripUpgrade(text: string): CantripUpgradeEntry[] {
  const entries: CantripUpgradeEntry[] = [];
  for (const m of text.matchAll(/(\d+)\s*\((\d+d\d+)\)/g)) {
    const level = parseInt(m[1], 10);
    if (level === 5 || level === 11 || level === 17) {
      const typeMatch = text.match(/(\d+d\d+)\s+(acid|cold|fire|force|lightning|necrotic|poison|radiant|thunder)/i);
      const type = typeMatch ? typeMatch[2].charAt(0).toUpperCase() + typeMatch[2].slice(1).toLowerCase() : 'Unknown';
      entries.push({ atCharacterLevel: level as 5 | 11 | 17, damage: [{ dice: m[2], type }] });
    }
  }
  return entries;
}

function migrateSpell(old: Record<string, unknown>): Spell {
  const spell = { ...old } as Record<string, unknown>;
  if (typeof spell.description === 'string') spell.description = [spell.description as string];
  delete spell.upcast;
  if (spell.damage && typeof spell.damage === 'object') delete (spell.damage as Record<string, unknown>).higherLevel;
  if (spell.heal && typeof spell.heal === 'object') delete (spell.heal as Record<string, unknown>).higherLevel;
  return spell as unknown as Spell;
}

// ── SRD Filter ────────────────────────────────────────────────

function isSrdSpell(spell: DndDataSpell): boolean {
  const pub = (spell.publisher || '').toLowerCase();
  const book = (spell.book || '').toLowerCase();
  if (pub.includes('wizards')) return true;
  for (const s of ['player', 'free basic', 'essentials', 'monster manual', "player's handbook"]) {
    if (book.includes(s)) return true;
  }
  return false;
}

// ── Transform ──────────────────────────────────────────────────

function transformSpell(spell: DndDataSpell): Spell {
  const props = spell.properties || {};
  const desc = spell.description || '';
  const parsed = parseDescription(desc);

  const result: Spell = {
    id: slugify(spell.name || ''),
    name: spell.name || '',
    level: props.Level ?? 0,
    school: (props.School || 'Unknown').charAt(0).toUpperCase() + (props.School || 'unknown').slice(1).toLowerCase(),
    castingTime: parseCastingTime(props['Casting Time']),
    range: props.Range || props['data-RangeAoe'] || 'Self',
    components: parseComponents(props.Components),
    duration: props.Duration || 'Instantaneous',
    concentration: /concentration/i.test(props.Duration || ''),
    ritual: checkRitual(props['Casting Time'] || '', desc),
    description: parsed.description,
    source: spell.book || spell.publisher || 'Unknown'
  };

  if (parsed.cantripUpgrade) result.cantripUpgrade = parsed.cantripUpgrade;
  if (parsed.usingAHigherLevelSpellSlot) result.usingAHigherLevelSpellSlot = parsed.usingAHigherLevelSpellSlot;

  const damage = extractDamage(desc);
  const heal = extractHeal(desc);
  const save = extractSave(desc);
  const attack = checkAttack(desc);

  if (damage) result.damage = damage;
  if (heal) result.heal = heal;
  if (save) result.save = save;
  if (attack) result.attack = attack;

  return result;
}

// ── Main ──────────────────────────────────────────────────────

async function main(): Promise<void> {
  console.log('Fetching dnd-data spells...');
  const url = 'https://raw.githubusercontent.com/nick-aschenbach/dnd-data/main/data/spells.json';
  let dndSpells: DndDataSpell[];
  try {
    const res = await fetch(url);
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    dndSpells = await res.json() as DndDataSpell[];
  } catch (err) {
    console.error('Failed to fetch:', err);
    process.exit(1);
  }

  console.log(`Total: ${dndSpells.length}`);
  const srdSpells = dndSpells.filter(isSrdSpell);
  console.log(`SRD: ${srdSpells.length}`);

  const transformed = srdSpells.map(s => transformSpell(s));
  console.log(`Transformed: ${transformed.length}`);

  // Load existing
  let existing: Spell[] = [];
  if (existsSync(localSpellsPath)) {
    existing = JSON.parse(readFileSync(localSpellsPath, 'utf-8')) as Spell[];
    console.log(`Existing: ${existing.length}`);
  }

  // Merge: prefer transformed for SRD spells
  const merged = new Map<string, Spell>();
  for (const s of transformed) merged.set(s.id, s);
  for (const s of existing) {
    if (!merged.has(s.id)) merged.set(s.id, migrateSpell(s as unknown as Record<string, unknown>));
  }

  // Sort
  const allSpells = Array.from(merged.values()).sort((a, b) => a.level !== b.level ? a.level - b.level : a.name.localeCompare(b.name));

  writeFileSync(localSpellsPath, JSON.stringify(allSpells, null, 2), 'utf-8');
  console.log(`Saved ${allSpells.length} spells to ${localSpellsPath}`);
}

main().catch(err => { console.error(err); process.exit(1); });
